crates/red_knot_workspace/src/workspace/metadata.rs

```diff
@@ -86,3 +237,600 @@ impl PackageMetadata {
+        }
+
+        let package = PackageMetadata::from_pyproject(pyproject, member_path, base_configuration)?;
+
+        tracing::debug!(
+            "Adding package '{}' at '{}'",
+            package.name(),
+            package.root()
+        );
+
+        packages.push(package);
+    }
+
+    Ok(CollectedPackagesOrStandalone::Packages(packages))
+}
+
+enum CollectedPackagesOrStandalone {
+    Packages(Vec<PackageMetadata>),
+    Standalone(PackageMetadata),
+}
+
+#[derive(Debug, Error)]
+pub enum WorkspaceDiscoveryError {
+    #[error("workspace path '{0}' is not a directory")]
+    NotADirectory(SystemPathBuf),
+
+    #[error("nested workspaces aren't supported but the package located at '{package_root}' defines a `knot.workspace` table")]
+    NestedWorkspaces { package_root: SystemPathBuf },
+
+    #[error("the workspace contains two packages named '{name}': '{first}' and '{second}'")]
+    DuplicatePackageNames {
+        name: Name,
+        first: SystemPathBuf,
+        second: SystemPathBuf,
+    },
+
+    #[error("the package '{package_name}' located at '{package_root}' is outside the workspace's root directory '{workspace_root}'")]
+    PackageOutsideWorkspace {
+        workspace_root: SystemPathBuf,
+        package_name: Name,
+        package_root: SystemPathBuf,
+    },
+
+    #[error(
+        "failed to read the `pyproject.toml` for the package located at '{package_root}': {source}"
+    )]
+    MemberFailedToReadPyProject {
+        package_root: SystemPathBuf,
+        source: std::io::Error,
+    },
+
+    #[error("{path} is not a valid `pyproject.toml`: {source}")]
+    InvalidPyProject {
+        source: Box<PyProjectError>,
+        path: SystemPathBuf,
+    },
+
+    #[error("invalid glob '{raw_glob}' in `tool.knot.workspace.members`: {source}")]
+    InvalidMembersPattern {
+        source: glob::PatternError,
+        raw_glob: String,
+    },
+
+    #[error("failed to match member glob: {error}")]
+    FailedToMatchGlob {
+        #[from]
+        error: GlobError,
+    },
+
+    #[error("the `[project]` is missing in the `pyproject.toml` for package '{package_path}'`")]
+    MissingProjectTable { package_path: SystemPathBuf },
+
+    #[error("the `project.name` is missing in the `pyproject.toml` for package '{package_path}'`")]
+    MissingPackageName { package_path: SystemPathBuf },
+}
+
+#[cfg(test)]
+mod tests {
+    //! Integration tests for workspace discovery
+
+    use crate::snapshot_workspace;
+    use anyhow::Context;
+    use insta::assert_ron_snapshot;
+    use ruff_db::system::{SystemPathBuf, TestSystem};
+
+    use crate::workspace::{WorkspaceDiscoveryError, WorkspaceMetadata};
+
+    #[test]
+    fn package_without_pyproject() -> anyhow::Result<()> {
+        let system = TestSystem::default();
+        let root = SystemPathBuf::from("/app");
+
+        system
+            .memory_file_system()
+            .write_files([(root.join("src/foo.py"), ""), (root.join("src/bar.py"), "")])
```

Comment:
> This is also related (in the opposite direction) to whether if they run red-knot on /app/src/foo/bar/ and /app/src/foo/__init__.py exists, should we walk up to the first directory that doesn't have __init__.py and determine that app/src/ is the actual root, or should we just let their imports fail because we are assuming the wrong root. (I think in this case we should do the walk-up.)

This is an interesting example and I think it might make sense for us to walk up the tree, but I'm not sure if it should change the workspace-root or only the inferred `SearchPathSettings`. I'd like to exclude this for now but it's something I should bring up in the CLI design document

Reply:
Yes, I think it's totally fine to defer the walking-up thing, as long as we have it noted somehow as something we probably want to address in future.

In general I don't think my comments on this PR are blocking, which is why I accepted. I don't think this PR does anything wrong, it's more that the rename to `workspace_root` and the way the projects in the tests are structured highlight an existing issue that I think we need to come to a resolution on.

Reply:
I went ahead and deleted all `src` usages 😆 Now there's no files left ;)